src/realizations/catchment/Bmi_Cpp_Formulation.cpp

```diff
@@ -0,0 +1,196 @@
+            }
+        }
+    }
+
+    while (next_time_step_index <= t_index) {
+        double model_initial_time = get_bmi_model()->GetCurrentTime();
+        set_model_inputs_prior_to_update(model_initial_time, t_delta);
+        if (t_delta_model_units == get_bmi_model()->GetTimeStep())
+            get_bmi_model()->Update();
+        else
+            get_bmi_model()->UpdateUntil(model_initial_time + t_delta_model_units);
+        // TODO: again, consider whether we should store any historic response, ts_delta, or other var values
+        next_time_step_index++;
+    }
+    return get_var_value_as_double( get_bmi_main_output_var());
+}
+
+double Bmi_Cpp_Formulation::get_var_value_as_double(const std::string& var_name) {
+    return get_var_value_as_double(0, var_name);
+}
+
+double Bmi_Cpp_Formulation::get_var_value_as_double(const int& index, const std::string& var_name) {
+    // TODO: consider different way of handling (and how to document) cases like long double or unsigned long long that
+    //  don't fit or might convert inappropriately
+    std::string type = get_bmi_model()->GetVarType(var_name);
+    if (type == "long double")
+        return (double) (get_bmi_model()->GetValuePtr<long double>(var_name))[index];
+
+    if (type == "double")
+        return (double) (get_bmi_model()->GetValuePtr<double>(var_name))[index];
+
+    if (type == "float")
+        return (double) (get_bmi_model()->GetValuePtr<float>(var_name))[index];
+
+    if (type == "short" || type == "short int" || type == "signed short" || type == "signed short int")
+        return (double) (get_bmi_model()->GetValuePtr<short>(var_name))[index];
+
+    if (type == "unsigned short" || type == "unsigned short int")
+        return (double) (get_bmi_model()->GetValuePtr<unsigned short>(var_name))[index];
+
+    if (type == "int" || type == "signed" || type == "signed int")
+        return (double) (get_bmi_model()->GetValuePtr<int>(var_name))[index];
+
+    if (type == "unsigned" || type == "unsigned int")
+        return (double) (get_bmi_model()->GetValuePtr<unsigned int>(var_name))[index];
+
+    if (type == "long" || type == "long int" || type == "signed long" || type == "signed long int")
+        return (double) (get_bmi_model()->GetValuePtr<long>(var_name))[index];
+
+    if (type == "unsigned long" || type == "unsigned long int")
+        return (double) (get_bmi_model()->GetValuePtr<unsigned long>(var_name))[index];
+
+    if (type == "long long" || type == "long long int" || type == "signed long long" || type == "signed long long int")
+        return (double) (get_bmi_model()->GetValuePtr<long long>(var_name))[index];
+
+    if (type == "unsigned long long" || type == "unsigned long long int")
+        return (double) (get_bmi_model()->GetValuePtr<unsigned long long>(var_name))[index];
+
+    throw std::runtime_error("Unable to get value of variable " + var_name + " from " + get_model_type_name() +
```

Comment:
Point taken, but this sounds like implementing our own stack trace functionality...is that not something we could extract from the exception already when we dump it to...wherever? (STDERR, file...)

Reply:
Aaaaand there I go again assuming standard C++ has functionality I've gotten used to in other languages.

Off-hand... using [`boost::stacktrace::stacktrace()`](https://www.boost.org/doc/libs/1_72_0/doc/html/stacktrace/getting_started.html) looks very interesting to me. I think we'd stand to get a lot more useful information for debugging by using that than trying to code-in potentially mutable names/details into our strings.

Reply:
@hellkite500, @mattw-nws, FWIW, I personally haven't generally included class names in log/error messages, mostly because of the relative difficulty with introspection in C++.   Without a dynamic runtime mechanism for gathering the name, I'd rather not go that route.   

Hard-coded strings would eventually not be maintained, not just failing, but undermining the intention of providing more clarity.  Also, it may not always be terribly helpful by itself to know the simple class name; e.g., if the exception happened in an super class function invoked by a subclass object.  And what about duplicate class names and namespaces?  

I think trying to keep individual exception messages unique somehow in their wording is a good practice.  But I don't see it as worth it to try to have a formal structure for such things, without also being able to leverage something to avoid or reduce manual maintenance of what has been applied.

As for @mattw-nws's suggestion, at first glance it looks like it could be useful.  But it seems to be something that could print stack trace information at the point where the framework is now throwing exceptions, not necessarily include more detail within the exception.  Considering adopting that approach is a bit of a different conversation, though perhaps one worth having.

Reply:
I think at some point this becomes a question of utility vs maintainability.

>FWIW, I personally haven't generally included class names in log/error messages, mostly because of the relative difficulty with introspection in C++. Without a dynamic runtime mechanism for gathering the name, I'd rather not go that route.

In lieu of having a way to quickly find the code related to an error/warning, detailed context on the error and states contributing to it would be a reasonable substitute (though that also comes with some maintainability issues I suppose.). The whole point of this original comment was that we have some good diagnostic/error checking in places, but when those events are triggered, it isn't always intuitive or easy to figure out what went wrong and how to resolve it.

I personally have had to do searches on the code base for error message strings, and then when there is more than one place a similar or the same string can generate from, I'm left up to much more manual debugging just to know where to start looking.

Reply:
Hey, can we bump our C++ version to C++20 to use [std::source_location](https://stackoverflow.com/a/58556021/489116)?

But seriously...

For this need it sounds like using `__FILE__` and `__LINE__` in exception messages would be a big help...you'd have to use them everywhere but it's fairly easy, no?

Reply:
See #388

Reply:
Also [58669c0](https://github.com/NOAA-OWP/ngen/pull/343/commits/58669c0c9ad3362372bb0ec899e056936cbd5d13)